test/README.md

```diff
@@ -1,22 +1,119 @@
 
-1. Unit tests
+1. Unit, Functional, and lint tests
 
-2. Manually triggered functional tests (without any Istio components)
+1. Istio e2e tests triggered on a istio/cni change
 
-3. Istio e2e tests triggered on a istio/cni change
+1. Istio e2e tests with the CNI enabled triggered on a istio/istio change
 
-4. Istio e2e tests with the CNI enabled triggered on a istio/istio change
+1. Manually executed e2e tests
 
-5. Manually executed e2e tests
+## Test Environments
 
-## Manually executed e2e tests
+The primary test environments for testing the Istio CNI plugin are:
+
+1. CircleCI
+1. Istio Prow (GKE)
+1. Developer private test environments
+
+### CircleCI Testing
+
+The "istio/cni" CircleCI test environment is under the CircleCI Istio organization.  It
+consists of 2 workflows:
+
+1. all:  PR events on the "istio/cni" github repository
+1. nightly:  triggers every night at 09:45:00 UTC
+
+All CI tests are run by both workflows.  The tests include the unit, functional, and
+lint tests as well as an e2e test.
+
+#### CircleCI e2e Tests
+
+The Istio CNI e2e_dind CircleCI job is patterned on the KDC CircleCI jobs.  The
+e2e_dind job creates a 3-node Kubernetes cluster via docker-in-docker (DinD)
+using the [kubeadm-dind-cluster (KDC)](https://github.com/kubernetes-sigs/kubeadm-dind-cluster)
+project.  Each node in the Kubernetes cluster is actually a docker container
+hosted on the CircleCI job's "remote" docker instance.  In order to run the
+Istio e2e_simple tests against that environment, a e2e_runner container is
+run within the same network namespace as the kubernetes master node
+container.  The e2e_runner image is based on the Istio circleci image containing
+golang, helm, kubectl, the istio/istio repo, and a simple script that runs
+the e2e_simple job with the Istio CNI enabled.
+
+The following illustrates the CircleCI job container relationship with the
+"remote" docker containers--kubernetes on DinD cluster containers and the
+e2e_runner container.
+
+![CircleCI e2e_dind job](images/istio_cni_circleci_e2e_test.svg)
+
+The "e2e_dind" job flow is as follows:
+
+1. The build job creates the docker image for the Istio CNI container: "install-cni".
+
+1. The build job persists the "install-cni" docker image for other jobs in the workspace.
+
+1. The e2e_dind job creates the e2e_runner image with the latest istio/istio content.
+
+1. The e2e_dind job brings up a DinD kubernetes cluster with flannel as the main CNI plugin type.
+
+1. The e2e_dind job populates the docker cache with the persisted "install-cni" image
+   for each kubernetes node.
+
+1. The e2e_dind job runs the e2e_runner container on the same docker network as the
+   kubernetes cluster's kube-master node.
+
+   1. The e2e_simple test is run by the e2e_runner exec script.
+
+##### kubeadm-dind-cluster (KDC) Comments
+
+- Implementations for multiple Kubernetes versions--since 1.8
+- Supports multiple CNI plugin types--weave, flannel, calico, calico_kdd, kube_router
```

Comment:
Not 100% sure, but I think most of these need to be capitalized and there are multiple instances through out the file

Reply:
I'm holding off on this as these are type values in the KDC config and this doc is developer/tester focused so don't want to get too pedantic here.

Reply:
If they are types then that is fine,  but you run amok of copyright/trademark if you don't properly capitalize things that should be capitalized.